roboteam_ai/src/coach/OffensiveCoach.cpp

```diff
@@ -0,0 +1,250 @@
+    }
+    drawOffensivePoints();
+}
+
+void OffensiveCoach::compareToCurrentPositions(const OffensiveCoach::OffensivePosition &position) {
+    OffensivePosition bestPos = position;
+    auto it = offensivePositions.begin();
+    while (it < offensivePositions.end()) {
+        if ((position.position - it->position).length() < Constants::OFFENSIVE_POSITION_DISTANCE()) {//Constants::ATTACKER_DISTANCE()) {
+            if (it->score >= bestPos.score) {
+                bestPos.position = it->position;
+                bestPos.score = it->score;
+            }
+            offensivePositions.erase(it);
+        }
+        it++;
+    }
+    offensivePositions.push_back(bestPos);
+}
+
+bool OffensiveCoach::compareByScore(OffensivePosition position1, OffensivePosition position2) {
+    return position1.score > position2.score;
+}
+
+Vector2 OffensiveCoach::calculatePositionForRobot(std::shared_ptr<roboteam_msgs::WorldRobot> robot) {
+    if ((robotPositions.find(robot->id) == robotPositions.end())) { // not there yet
+        return getClosestOffensivePosition(robot);
+
+    } else {
+        calculateNewRobotPositions(robot);
+        if ((robotPositions.find(robot->id) == robotPositions.end())) { //not in there
+            return calculatePositionForRobot(robot);
+        } else {
+            return robotPositions[robot->id].position;
+        }
+    }
+}
+
+Vector2 OffensiveCoach::getClosestOffensivePosition(const shared_ptr<roboteam_msgs::WorldRobot> &robot) {
+    double distance = INT_MAX;
+    double currentDistance;
+    OffensivePosition newRobotPosition;
+    for (auto &position : offensivePositions) {
+        currentDistance = (position.position - robot->pos).length();
+        if (currentDistance < distance) {
+            distance = currentDistance;
+            newRobotPosition = position;
+        }
+    }
+    if (distance == INT_MAX) {
+        newRobotPosition.position = robot->pos;
+        newRobotPosition.score = CoachHeuristics::calculatePositionScore(robot->pos);
+    }
+
+
+    if ((newRobotPosition.position - robot->pos).length() < 0.8) {
+        robotPositions[robot->id] = newRobotPosition;
+    }
+    return newRobotPosition.position;
+}
+
+void OffensiveCoach::releaseRobot(int robotID) {
+    robotPositions.erase(robotID);
+}
+
+Vector2 OffensiveCoach::getPositionForRobotID(int robotID) {
+    Vector2 position = robotPositions[robotID].position;
+    return position;
+}
+
+void OffensiveCoach::calculateNewRobotPositions(std::shared_ptr<roboteam_msgs::WorldRobot> robot) {
+    OffensiveCoach::OffensivePosition currentRobotPosition = robotPositions[robot->id];
+
+    Vector2 currentPosition = currentRobotPosition.position;
+    double currentScore = CoachHeuristics::calculatePositionScore(currentPosition);
+
+    OffensivePosition newPosition;
+    bool foundNewPosition = false;
+
+    for (int xDiff : {-2, -1, 0, 1, 2}) {
```

Comment:
this looks pretty ugly

Reply:
Yeah I know, but would not know how to make it much nicer

Reply:
Nvm, made it nicer